### DISCUSSION TITLE
Quantizing models with bitsandbytes

### DISCUSSION THREAD

**Discussion by raunaks13** [bitsandbytes-foundation/bitsandbytes#1163] [Q&A]
I had a couple of questions related to exactly how the quantization is done:

1. Does bitsandbytes do fp16 -> int8 quantization after transferring the tensors to the GPU? And if you want to dequantize, are those operations done on the GPU as well?
2. I traced the workflow of of `Linear8BitLt()`, which leads me to believe that quantization is happening in this line - https://github.com/TimDettmers/bitsandbytes/blob/main/csrc/kernels.cu#L2419. Could someone please confirm this? If not, where is the quantization occuring?
3. Is the quantization method used absmax or zero point? Is it done row-wise? There is some mention of column-wise features, but when I load quantized models with huggingface the scale factors seem to be different for each row, not column.
4. When you quantize a model, do you treat outliers separately as described in the LLM.int8() paper? If so, then where does this happen in the source code?

**Comment by gyanu2507** [comment]
1. Weights are quantized on the GPU when the 8-bit Linear is filled (`Linear8bitLt`). Dequant + matmul is a fused CUDA kernel, also on GPU. You do not round-trip to CPU for the matmul.

2. The line you found in `kernels.cu` is in that fused path (INT8 GEMM + dequant). The conversion of fp16/fp32 weights into the INT8 storage happens at module init / `quantize` time, not on every forward. Follow `Int8Params` / `quantize_blockwise` rather than the inner GEMM loop if you want the cast itself.

3. LLM.int8() is absmax, not zero-point. Scales are per-vector. Hugging Face showing a scale per row is expected: that is the output-feature / row dimension of the weight. The paper also talks about column-wise outlier detection on *activations*, which is a different tensor.

4. Outliers: yes, that is the whole point of LLM.int8(). Hidden states with large magnitude are extracted and multiplied in fp16. In the current tree that lives in the 8-bit linear forward (threshold ~6.0 by default, `threshold` on `Linear8bitLt`). Weights themselves stay INT8 plus the fp16 outlier state.

`load_in_8bit` via transformers uses this path. NF4 / fp4 is a different quantizer (`Linear4bit`) and does not use the same outlier split.
